### PR TITLE
[Tabs] Adds justified and scrollable layout to MDCTabBarView

### PR DESCRIPTION
**Pull request by leonmz**
Adds justified and scrollable layout for MDCTabBarView and corresponding tests.

## Example View Controller

|Before|After (Portrait)|After (Landscape)
|---|---|---|
|60151483-ea246900-97aa-11e9-9c1e-688f1d241512|Simulator Screen Shot - iPhone 7 - 2019-06-26 at 08 27 47|Simulator Screen Shot - iPhone 7 - 2019-06-26 at 08 29 09|

Part of Issue #7645

### CODE REVIEW

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 1]
```diff
@@ -15,13 +15,15 @@
 #import "MDCTabBarView.h"
 #import "private/MDCTabBarViewItemView.h"
 
+#import "private/MDCTabBarViewItemView.h"
```
Comment (mikefan1991) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: This is duplicate import.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 2]
```diff
@@ -30,10 +32,10 @@ @implementation MDCTabBarView
 #pragma mark - Initialization
 
 - (instancetype)init {
-  self = [super init];
+  self = [super initWithFrame:CGRectZero];
```
Comment (mikefan1991) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: This change is not necessary.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 3]
```diff
@@ -93,28 +80,62 @@ - (void)setSelectedItem:(UITabBarItem *)selectedItem {
...
+  for (UIView *view in self.stackView.arrangedSubviews) {
+    [view removeFromSuperview];
+    [_stackView removeArrangedSubview:view];
+  }
+
+  for (UITabBarItem *item in self.items) {
+    MDCTabBarViewItemView *itemView = [[MDCTabBarViewItemView alloc] init];
+    itemView.translatesAutoresizingMaskIntoConstraints = NO;
+    itemView.title = item.title;
+    itemView.image = item.image;
```
Comment (mikefan1991) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: I have removed the title and the image in another PR. Please wait for that to be merged and then update this.

File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m [hunk 4]
```diff
@@ -48,7 +48,7 @@ - (void)setUp {
 
   // Uncomment below to recreate all the goldens (or add the following line to the specific
   // test you wish to recreate the golden for).
-  //  self.recordMode = YES;
+  self.recordMode = YES;
```
Comment (mikefan1991) [inline components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m]: This should stay commended.

**Review by mikefan1991 — CHANGES_REQUESTED** [review]

**Comment by romoore** [comment]
Gonna take a stab at cleaning this up. Ping me if you plan to work on it as well.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 5]
```diff
@@ -30,10 +32,10 @@ @implementation MDCTabBarView
 #pragma mark - Initialization
 
 - (instancetype)init {
-  self = [super init];
+  self = [super initWithFrame:CGRectZero];
   if (self) {
     _items = @[];
-    _itemViews = @[];
+    [self setUpSubviews];
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: This looks far too complex to be called from `-init`.  Let's try to simplify and add the constraints later.

**Review by romoore — CHANGES_REQUESTED** [review]

File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m (hunk 4 above)
Comment (romoore) [reply components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m]: Done.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 3 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Done.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 2 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Reverted.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 1 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Removed.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 6]
```diff
@@ -20,8 +20,8 @@
 
 @interface MDCTabBarView ()
 
-/** The views representing the items of this tab bar. */
-@property(nonatomic, strong) NSArray<UIView *> *itemViews;
+/** The stack view that contains all tab items. */
+@property(nonatomic) UIStackView *stackView;
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Nit: Please document retention.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 7]
```diff
@@ -93,28 +78,62 @@ - (void)setSelectedItem:(UITabBarItem *)selectedItem {
   _selectedItem = selectedItem;
 }
 
-#pragma mark - UIView
+- (void)setUpSubviews {
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Please rename this to `commonMDCTabBarViewInit`.

**Review by romoore — CHANGES_REQUESTED** [review]

File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m [hunk 8]
```diff
@@ -137,4 +137,63 @@ - (void)testItemsWithMixedTitlesAndImages {
   [self generateSnapshotAndVerifyForView:self.tabBarView];
 }
 
+- (void)testMultipleLineItemViewToFitInTheSingleScreen {
```
Comment (romoore) [inline components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m]: Please add a similar test that uses the bar's `intrinsicContentSize` to set the bounds so we can be sure it will return the correct size to fit the items.

File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m [hunk 9]
```diff
@@ -137,4 +137,63 @@ - (void)testItemsWithMixedTitlesAndImages {
...
+  UITabBarItem *item6 = [[UITabBarItem alloc] initWithTitle:@"Six" image:nil tag:8];
+
+  // When
+  self.tabBarView.items = @[ item1, item2, item3, item4, item5, item6 ];
+
+  // Then
+  [self generateSnapshotAndVerifyForView:self.tabBarView];
+}
+
+- (void)testTooManyItemsInScrollView {
```
Comment (romoore) [inline components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m]: Nit: Naming: Please remove "InScollView" from the test since the behavior is automatic and depends on bounds sizes, etc.  Consider the name `testItemsLargerThanBoundsChangesToScrollableLayout`?

File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m [hunk 10]
```diff
@@ -137,4 +137,63 @@ - (void)testItemsWithMixedTitlesAndImages {
...
+  UITabBarItem *item8 = [[UITabBarItem alloc] initWithTitle:@"Eight" image:nil tag:8];
+
+  // When
+  self.tabBarView.items = @[ item1, item2, item3, item4, item5, item6, item7, item8 ];
+
+  // Then
+  [self generateSnapshotAndVerifyForView:self.tabBarView];
+}
+
+- (void)testSuperLongItemViewInScrollView {
```
Comment (romoore) [inline components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m]: A few things here: 
*  Please make the bounds larger than the single item's maximum width (I believe it's 360)
*  Please remove "InScrollView" from the name, since the layout depends on content + bounds and isn't guaranteed.  Perhaps `testVeryLongItemLimitsWidth` or something similar?

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 6 above)
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Not sure if this change resolve this issue..

File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m (hunk 8 above)
Comment (leonmz) [reply components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m]: Added.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 5 above)
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Fixed.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 7 above)
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Fix applied.

File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m (hunk 9 above)
Comment (leonmz) [reply components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m]: Thanks for the advising!

File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m (hunk 10 above)
Comment (leonmz) [reply components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m]: Added :)

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 6 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: I don't see a difference. Perhaps you need to push? I think it's also better to name this for its semantic use (rather than its type).  I think something like this would be better:

```objc
@property(nonnull, nonatomic, strong) UIStackView *containerView;
```

Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Fix applied!

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 11]
```diff
@@ -99,32 +87,88 @@ - (void)setSelectedItem:(UITabBarItem *)selectedItem {
   _selectedItem = selectedItem;
 }
 
-#pragma mark - UIView
```
Comment (mikefan1991) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: #pragma mark - Private API should be added for the following functions.

Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Fixed.

**Review by mikefan1991 — APPROVED** [review]

**Comment by leonmz** [comment]
I refactored the colorFortState function a bit. Deleted the items view and use the arranged view from containerView instead. Please take a look at it. Thanks!

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 12]
```diff
@@ -20,8 +20,11 @@
 
 @interface MDCTabBarView ()
 
-/** The views representing the items of this tab bar. */
-@property(nonnull, nonatomic, strong) NSArray<UIView *> *itemViews;
+/** The stack view that contains all tab item views. */
+@property(nonnull, nonatomic, strong) UIStackView *containerView;
+
+/** The constraints managing this view. */
+@property(nonatomic, strong) NSArray *viewConstraints;
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Please add a nullability annotation to this property. Please also use generics to document that it contains `NSLayoutConstraint *` types.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 13]
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
...
+  NSMutableArray *constraints = [NSMutableArray array];
+  [constraints addObjectsFromArray:@[
+    [_containerView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
+    [_containerView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
+    [_containerView.widthAnchor constraintGreaterThanOrEqualToAnchor:self.widthAnchor],
+    [_containerView.topAnchor constraintEqualToAnchor:self.topAnchor],
+    [_containerView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor],
+  ]];
+
+  [self addConstraints:constraints];
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Please don't call `addConstraints` since it will be deprecated. Instead, activate the constraints directly.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 14]
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
...
+    CGSize contentSize = itemView.intrinsicContentSize;
+    if (contentSize.width > maxWidth) {
+      maxWidth = contentSize.width;
+    }
   }
-  CGSize boundsSize = CGSizeMake(CGRectGetWidth(self.bounds), CGRectGetHeight(self.bounds));
-  CGSize itemSize = CGSizeMake(boundsSize.width / self.items.count, boundsSize.height);
+  CGFloat requiredWidth = maxWidth * self.items.count;
+  BOOL canBeJustified = availableWidth > requiredWidth;
+  self.containerView.distribution = canBeJustified ? UIStackViewDistributionFillEqually
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m:275]: It seems like `layoutSubviews` shouldn't be required if a constraint-based layout is used. Even then, I would expect that this code execute before `[super layoutSubviews]` since that is expected to use any constraints to layout the subviews.

I suspect a better place to configure `self.containerView.distribution` might be in `setItems` and in `setBounds`.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 15]
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
...
+  _containerView = [[UIStackView alloc] init];
+  _containerView.axis = UILayoutConstraintAxisHorizontal;
+  _containerView.translatesAutoresizingMaskIntoConstraints = NO;
+  [self addSubview:_containerView];
+}
+
+- (void)setUpItemViews {
+  for (UIView *view in self.containerView.arrangedSubviews) {
+    [view removeFromSuperview];
+    [_containerView removeArrangedSubview:view];
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: This isn't necessary since `removeFromSuperview` automatically removes it from the stack view's `arrangedSubviews` property.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 16]
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
...
 #pragma mark - UIView
 
-// TODO(#7645): Temporary layout until
-// https://github.com/material-components/material-components-ios/issues/7645 lands
 - (void)layoutSubviews {
   [super layoutSubviews];
 
-  if (self.items.count == 0) {
-    return;
+  CGFloat availableWidth = self.bounds.size.width;
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Nit: Please use `CGRectGetWidth(self.bounds)` instead of accessing the bounds directly. Especially for UIScrollViews, the bounds origin may not be (0, 0).

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 17]
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
...
+  CGFloat maxWidth = 0;
+  for (UIView *itemView in self.containerView.arrangedSubviews) {
+    CGSize contentSize = itemView.intrinsicContentSize;
+    if (contentSize.width > maxWidth) {
+      maxWidth = contentSize.width;
+    }
   }
-  CGSize boundsSize = CGSizeMake(CGRectGetWidth(self.bounds), CGRectGetHeight(self.bounds));
-  CGSize itemSize = CGSizeMake(boundsSize.width / self.items.count, boundsSize.height);
+  CGFloat requiredWidth = maxWidth * self.items.count;
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Looks like this can be replaced with a call to `self.intrinsicContentSize`?

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 18]
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
...
+      maxHeight = contentSize.height;
+    }
+  }
+  return CGSizeMake(totalWidth, MAX(kMinHeight, maxHeight));
 }
 
 - (CGSize)sizeThatFits:(CGSize)size {
-  return CGSizeMake(size.width, MAX(size.height, kMinHeight));
+  CGSize intrinsicSize = self.intrinsicContentSize;
+  return CGSizeMake(MIN(intrinsicSize.width, size.width), MAX(intrinsicSize.height, size.height));
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: I'm curious about this logic. If the current bounds is `CGRectZero` and `sizeToFit` is called, then this view would only resize its height. Is that expected?  Should it not attempt to provide a fit size for its contents as a minimum, but allow a larger width?

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 19]
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
...
+}
+
+#pragma mark - Private Functions
+
+- (void)commonMDCTabBarViewInit {
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Nit: Please move this up near the `-init` methods.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 20]
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
...
+  self.backgroundColor = UIColor.whiteColor;
+
+  [self setUpStackView];
+  [self setUpItemViews];
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: This call doesn't seem necessary.  I believe the `items` array is empty during initialization.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 21]
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
...
+}
+
+- (void)setUpStackView {
+  _containerView = [[UIStackView alloc] init];
+  _containerView.axis = UILayoutConstraintAxisHorizontal;
+  _containerView.translatesAutoresizingMaskIntoConstraints = NO;
+  [self addSubview:_containerView];
+}
+
+- (void)setUpItemViews {
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: I believe this logic can be put back in `setItems:` since it's only called from there.

**Review by romoore — CHANGES_REQUESTED** [review]
I'm going to address most of my feedback and push up changes to the branch.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 17 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Ah, no, that returns the scrollable width, not the justified width.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 14 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m:275]: I'll leave this here for now. `setBounds:` is called more times during the initial presentation of a view and not much less often during rotation. It's likely not worth the optimization.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 21 above)
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: I still feel it worth a function extract the logic..

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 12 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Done. I was able to remove this property and replace it with a flag. #namingIsHard

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 13 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Done

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 16 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Done.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 19 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Done. Moved into `-init`

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 20 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Removed.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 21 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Could you help me understand your thinking? We cannot unit test this code and it's only ever called from within `setItems:`. Either testability or duplicate code would be good candidates for a method. In this case I'm trying to understand the value of an extra message send (plus complexity in figuring out what the method does).

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 18 above)
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: I am thinking about since it is a scrollable view thus this should be fits in any width.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 22]
```diff
@@ -142,30 +146,56 @@ - (UIColor *)titleColorForState:(UIControlState)state {
...
+    [super updateConstraints];
+    return;
   }
+
+  [_containerView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor].active = YES;
+  [_containerView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor].active = YES;
+  [_containerView.widthAnchor constraintGreaterThanOrEqualToAnchor:self.widthAnchor].active = YES;
+  [_containerView.topAnchor constraintEqualToAnchor:self.topAnchor].active = YES;
+  [_containerView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor].active = YES;
+  self.containerViewConstraintsActive = YES;
```
[the remark this answers is not in the document]
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m:291]: Done?

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 15 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Removed.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 18 above)
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: CGZero is a extremely case, but if a small bounds is provided, the scrollable view should also fit itself in

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 21 above)
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Thanks  a lot for detailed explanation!!

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 18 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: That's true, but the fit size would (ideally) be the size of its content. I think the best thing to conform to the documented behavior of `sizeThatFits:` is to return the "fit size".  In this case, it's probably the the size of the justified layout (rather than the intrinsic content size).  Does that make sense?

Whether it's kept like this or changed, I think it's worth adding a unit test that calls `sizeToFit:` with a bounds of `CGRectZero` and make sure it gets back a non-zero width and non-zero height.  Something similar for a starting bounds bigger than necessary?

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 23]
```diff
@@ -142,30 +147,56 @@ - (UIColor *)titleColorForState:(UIControlState)state {
...
+
+  // Must always be called last according to the documentation.
+  [super updateConstraints];
 }
 
 - (CGSize)intrinsicContentSize {
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m:297]: Please add unit tests to cover verify the behavior/values of `intrinsicContentSize`.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 24]
```diff
@@ -142,30 +147,56 @@ - (UIColor *)titleColorForState:(UIControlState)state {
...
+    CGSize contentSize = itemView.intrinsicContentSize;
+    totalWidth += contentSize.width;
+    if (contentSize.height > maxHeight) {
+      maxHeight = contentSize.height;
+    }
+  }
+  return CGSizeMake(totalWidth, MAX(kMinHeight, maxHeight));
 }
 
 - (CGSize)sizeThatFits:(CGSize)size {
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m:309]: I also mentioned it in another thread, but please add unit tests for `sizeThatFits:`

**Review by romoore — CHANGES_REQUESTED** [review]
Looks good, but needs unit tests for `intrinsicContentSize` and `sizeThatFits:` before it's ready for merge.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 23 above)
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m:297]: Added.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 24 above)
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m:309]: Added.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 18 above)
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Sure. Fixed applied. I also refactored the intrisicSize to justified width instead of scrollable width.

**Comment by leonmz** [comment]
Refactored the intrsicSize to justified size. Also added snapshot tests for case that the bounds is 1000 but only two items exist. The current behavior is that the item view has been stretched to 500 due to its low hugging priority.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 23 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m:297]: No, I don't think the TabBarView should be dependent on its contents.  I'd prefer the tests cover the more general behavior of `intrinsicContentSize`. For example, ensuring that it returns the justified/scrollable width consistently (depending on what you determine is more useful). That the height grows if one of its items has a very large image or long text.  It's the relationship/behavior that's important and not the exact values.

Let me add one or two as a commit and we can talk about them.  I think that will be faster and clearer than my trying to explain in a comment. ;-)

**Review by romoore — APPROVED** [review]

**Comment by romoore** [comment]
@googlebot All authors are Googlers, but one commit has a misconfigured author email.
